micro-rdk-nmea/src/viamboat/sensors.rs

```diff
@@ -0,0 +1,579 @@
+            if split_key.len() == 2 {
+                if let Ok(msg_pgn) = u32::from_str_radix(split_key[0], 16) {
+                    pgns.iter().any(|&x| x == msg_pgn)
+                } else {
+                    false
+                }
+            } else {
+                false
+            }
+        } else {
+            true
+        }
+    });
+
+    let filtered_on_src = filtered_on_pgn.filter(|(k, _)| {
+        if let Some(sources) = &sources {
+            let split_key: Vec<&str> = k.split_terminator("-").collect();
+            if split_key.len() == 2 {
+                if let Ok(src) = split_key[1].parse::<u32>() {
+                    sources.iter().any(|&x| (x as u32) == src)
+                } else {
+                    false
+                }
+            } else {
+                false
+            }
+        } else {
+            true
+        }
+    });
+
+    let res: Result<Vec<NmeaMessage>, ViamboatSensorError> = filtered_on_src
+        .map(|(_, v)| {
+            let inner = v
+                .kind
+                .as_ref()
+                .ok_or(ViamboatSensorError::EmptyValueError)?;
+            match inner {
+                Kind::StringValue(value_str) => {
+                    let mut data = Vec::<u8>::new();
+                    general_purpose::STANDARD.decode_vec(value_str.as_str(), &mut data)?;
+                    Ok(NmeaMessage::try_from(data)?)
+                }
+                _ => Err(ViamboatSensorError::ValueTypeError),
+            }
+        })
+        .collect();
+
+    res.and_then(|msgs| {
+        if msgs.is_empty() {
+            Err(ViamboatSensorError::NoAvailableReadings)
+        } else {
+            Ok(msgs)
+        }
+    })
+}
+
+impl PgnGateway {
+    fn retrieve_messages(
+        &mut self,
+        pgns: Option<Vec<u32>>,
+        sources: Option<Vec<u8>>,
+    ) -> Result<Vec<NmeaMessage>, ViamboatSensorError> {
+        let all_pgns = self.sensor.get_generic_readings()?;
+        readings_to_messages(&all_pgns, pgns, sources)
+    }
+}
+
+pub fn get_raw_sensor_key(cfg: ConfigType) -> Vec<ResourceKey> {
+    let mut dep_keys = vec![];
+    if let Ok(sensor_name) = cfg.get_attribute::<String>("raw_pgn_sensor") {
+        dep_keys.push(ResourceKey::new(SensorCompName, sensor_name));
+    }
+    dep_keys
+}
+
+fn get_raw_sensor_dependency(
+    cfg: &ConfigType,
+    deps: Vec<Dependency>,
+) -> Result<SensorType, SensorError> {
+    let raw_pgn_sensor_name = cfg.get_attribute::<String>("raw_pgn_sensor")?;
+    let mut raw_pgn_sensor: Option<SensorType> = None;
+    for Dependency(k, v) in deps {
+        if let Resource::Sensor(found_sensor) = v {
+            if k.1 == raw_pgn_sensor_name {
+                let _ = raw_pgn_sensor.insert(found_sensor.clone());
+            }
+        }
+    }
+    raw_pgn_sensor.ok_or(SensorError::ConfigError("missing dependent raw pgn sensor"))
+}
+
+const DEPTH_PGN: u32 = 128267;
```

Comment:
Wouldn't it be better to get these PGN's out from the constants? Isn't this just `WaterDepth::PGN`, which is a `CONST` so should be useable here?

Reply:
Since those structs (and their names) will be auto-generated in the future, the true constant is the PGN